File: android/app/src/main/java/com/reunited/countdown/reunited_countdown/CountdownWidgetProvider.java
package com.reunited.countdown.reunited_countdown;

import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.widget.RemoteViews;

import java.text.SimpleDateFormat;
import java.util.Locale;
import java.util.TimeZone;

public class CountdownWidgetProvider extends AppWidgetProvider {

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        for (int appWidgetId : appWidgetIds) {
            updateWidget(context, appWidgetManager, appWidgetId);
        }
        startService(context);
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        super.onReceive(context, intent);
        if (Intent.ACTION_BOOT_COMPLETED.equals(intent.getAction())) {
            startService(context);
        }
    }

    @Override
    public void onEnabled(Context context) {
        startService(context);
    }

    // --- Service start ---

    static void startService(Context context) {
        Intent intent = new Intent(context, CountdownWidgetService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            context.startForegroundService(intent);
        } else {
            context.startService(intent);
        }
    }

    // --- Widget rendering ---

    static void updateWidget(Context context, AppWidgetManager appWidgetManager, int appWidgetId) {
        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.countdown_widget);
        SharedPreferences prefs = context.getSharedPreferences("FlutterSharedPreferences", Context.MODE_PRIVATE);

        String reunionDateStr = prefs.getString("flutter.reunionDate", null);
        String timezone = prefs.getString("flutter.selectedTimezone", "Indonesia");

        int days, hours, minutes, seconds;
        boolean isTimeUp;

        if (reunionDateStr != null) {
            long[] result = calculateCountdown(reunionDateStr, timezone);
            if (result == null) {
                days    = prefs.getInt("flutter.widget_days", -1);
                hours   = prefs.getInt("flutter.widget_hours", -1);
                minutes = prefs.getInt("flutter.widget_minutes", -1);
                seconds = prefs.getInt("flutter.widget_seconds", -1);
                isTimeUp = prefs.getBoolean("flutter.widget_isTimeUp", false);
            } else {
                long remainingSec = result[0];
                isTimeUp = remainingSec <= 0;
                days    = isTimeUp ? 0 : (int) (remainingSec / 86400);
                hours   = isTimeUp ? 0 : (int) ((remainingSec % 86400) / 3600);
                minutes = isTimeUp ? 0 : (int) ((remainingSec % 3600) / 60);
                seconds = isTimeUp ? 0 : (int) (remainingSec % 60);
            }
        } else {
            days    = prefs.getInt("flutter.widget_days", -1);
            hours   = prefs.getInt("flutter.widget_hours", -1);
            minutes = prefs.getInt("flutter.widget_minutes", -1);
            seconds = prefs.getInt("flutter.widget_seconds", -1);
            isTimeUp = prefs.getBoolean("flutter.widget_isTimeUp", false);
        }

        android.util.Log.d("CountdownWidget", "Read values - Days: " + days + ", Hours: " + hours + ", Minutes: " + minutes + ", Seconds: " + seconds);

        if (isTimeUp) {
            views.setTextViewText(R.id.widget_title, "🎉 C'est l'heure ! 🎉");
            views.setTextViewText(R.id.widget_days, "🎊");
            views.setTextViewText(R.id.widget_hours, "💕");
            views.setTextViewText(R.id.widget_minutes, "🎉");
            views.setTextViewText(R.id.widget_seconds, "💖");
        } else if (days == -1) {
            views.setTextViewText(R.id.widget_title, "💕 Reunited Countdown");
            views.setTextViewText(R.id.widget_days, "⏳");
            views.setTextViewText(R.id.widget_hours, "⏳");
            views.setTextViewText(R.id.widget_minutes, "⏳");
            views.setTextViewText(R.id.widget_seconds, "⏳");
        } else {
            views.setTextViewText(R.id.widget_title, "💕 Reunited Countdown");
            views.setTextViewText(R.id.widget_days,    String.format("%02d", days));
            views.setTextViewText(R.id.widget_hours,   String.format("%02d", hours));
            views.setTextViewText(R.id.widget_minutes, String.format("%02d", minutes));
            views.setTextViewText(R.id.widget_seconds, String.format("%02d", seconds));
        }

        Intent openApp = new Intent(context, MainActivity.class);
        android.app.PendingIntent openPi = android.app.PendingIntent.getActivity(context, 0, openApp,
                android.app.PendingIntent.FLAG_UPDATE_CURRENT | android.app.PendingIntent.FLAG_IMMUTABLE);
        views.setOnClickPendingIntent(R.id.widget_container, openPi);

        appWidgetManager.updateAppWidget(appWidgetId, views);
    }

    public static void updateAllWidgets(Context context) {
        AppWidgetManager manager = AppWidgetManager.getInstance(context);
        int[] ids = manager.getAppWidgetIds(new ComponentName(context, CountdownWidgetProvider.class));
        for (int id : ids) {
            updateWidget(context, manager, id);
        }
    }

    // --- Countdown calculation ---

    private static long[] calculateCountdown(String reunionDateStr, String timezone) {
        try {
            SimpleDateFormat sdf = tryParse(reunionDateStr);
            if (sdf == null) return null;
            sdf.setTimeZone(TimeZone.getTimeZone("UTC"));
            long reunionNaiveMs = sdf.parse(reunionDateStr).getTime();
            int destOffsetHours = getTimezoneOffset(timezone);
            long reunionUTC = reunionNaiveMs - (long) destOffsetHours * 3_600_000L;
            long remainingMs = reunionUTC - System.currentTimeMillis();
            return new long[]{ Math.max(0, remainingMs / 1000) };
        } catch (Exception e) {
            return null;
        }
    }

    private static SimpleDateFormat tryParse(String s) {
        if (s != null && s.length() >= 23 && s.charAt(19) == '.') {
            return new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS", Locale.US);
        } else if (s != null && s.length() >= 19) {
            return new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.US);
        }
        return null;
    }

    private static int getTimezoneOffset(String timezone) {
        if ("France".equals(timezone)) {
            return TimeZone.getTimeZone("Europe/Paris")
                    .getOffset(System.currentTimeMillis()) / 3_600_000;
        } else if ("Indonesia".equals(timezone)) {
            return 7;
        }
        return 0;
    }
}

File: android/app/src/main/kotlin/com/reunited/countdown/reunited_countdown/CountdownWidgetProvider.kt
package com.reunited.countdown.reunited_countdown

import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.widget.RemoteViews
import android.app.AlarmManager
import android.app.PendingIntent
import java.util.Calendar
import java.util.concurrent.TimeUnit

class CountdownWidgetProvider : AppWidgetProvider() {
    
    companion object {
        private const val ACTION_UPDATE_WIDGET = "com.reunited.countdown.reunited_countdown.UPDATE_WIDGET"
    }
    
    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        for (appWidgetId in appWidgetIds) {
            updateAppWidget(context, appWidgetManager, appWidgetId)
        }
        
        // Programmer la prochaine mise à jour
        scheduleNextUpdate(context)
    }
    
    private fun updateAppWidget(context: Context, appWidgetManager: AppWidgetManager, appWidgetId: Int) {
        // Calculer le temps restant jusqu'au 15 août 2025
        val target = Calendar.getInstance().apply {
            set(2025, Calendar.AUGUST, 15, 0, 0, 0)
            set(Calendar.MILLISECOND, 0)
        }
        
        val currentTime = System.currentTimeMillis()
        val targetTime = target.timeInMillis
        val timeLeft = targetTime - currentTime
        
        val countdownText: String
        val subtitle: String
        
        if (timeLeft > 0) {
            val days = TimeUnit.MILLISECONDS.toDays(timeLeft)
            val hours = TimeUnit.MILLISECONDS.toHours(timeLeft) % 24
            val minutes = TimeUnit.MILLISECONDS.toMinutes(timeLeft) % 60
            
            when {
                days > 0 -> {
                    countdownText = String.format("%d jours\n%02d:%02d", days, hours, minutes)
                    subtitle = "💕"
                }
                hours > 0 -> {
                    countdownText = String.format("%02d:%02d", hours, minutes)
                    subtitle = "Dernières heures ! 💕"
                }
                else -> {
                    countdownText = String.format("%d min", minutes)
                    subtitle = "Bientôt ! 💕"
                }
            }
        } else {
            countdownText = "C'est le jour !"
            subtitle = "Retrouvailles ! 🎉💕"
        }
        
        // Créer les vues du widget
        val views = RemoteViews(context.packageName, R.layout.widget_layout)
        views.setTextViewText(R.id.countdown_display, countdownText)
        views.setTextViewText(R.id.widget_title, "Retrouvailles")
        
        // Intent pour ouvrir l'app au clic
        val intent = Intent().apply {
            setClassName(context, "com.reunited.countdown.reunited_countdown.MainActivity")
        }
        val pendingIntent = PendingIntent.getActivity(
            context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
        views.setOnClickPendingIntent(R.id.widget_root, pendingIntent)
        
        // Mettre à jour le widget
        appWidgetManager.updateAppWidget(appWidgetId, views)
    }
    
    private fun scheduleNextUpdate(context: Context) {
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val intent = Intent(context, CountdownWidgetProvider::class.java).apply {
            action = ACTION_UPDATE_WIDGET
        }
        val pendingIntent = PendingIntent.getBroadcast(
            context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
        
        // Mettre à jour toutes les minutes
        val nextUpdate = System.currentTimeMillis() + 60000 // 1 minute
        alarmManager.setExact(AlarmManager.RTC, nextUpdate, pendingIntent)
    }
    
    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)
        
        if (ACTION_UPDATE_WIDGET == intent.action) {
            val appWidgetManager = AppWidgetManager.getInstance(context)
            val componentName = ComponentName(context, CountdownWidgetProvider::class.java)
            val appWidgetIds = appWidgetManager.getAppWidgetIds(componentName)
            onUpdate(context, appWidgetManager, appWidgetIds)
        }
    }
    
    override fun onEnabled(context: Context) {
        super.onEnabled(context)
        // Premier widget ajouté
        scheduleNextUpdate(context)
    }
    
    override fun onDisabled(context: Context) {
        super.onDisabled(context)
        // Dernier widget supprimé - annuler les mises à jour
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val intent = Intent(context, CountdownWidgetProvider::class.java).apply {
            action = ACTION_UPDATE_WIDGET
        }
        val pendingIntent = PendingIntent.getBroadcast(
            context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
        alarmManager.cancel(pendingIntent)
    }
}
